features/bsm/daemon/src/main/java/org/opennms/netmgt/bsm/daemon/Bsmd.java

```diff
@@ -196,7 +203,70 @@ public void serviceInterfaceOrNodeDeleted(Event e) {
                 || EventConstants.SERVICE_DELETED_EVENT_UEI.equals(e.getUei()) && reductionKeys.contains(String.format("uei.opennms.org/nodes/nodeLostService::%d:%s:%s", e.getNodeid(), e.getInterface(), e.getService()))
                 || EventConstants.APPLICATION_DELETED_EVENT_UEI.equals(e.getUei()) && m_stateMachine.getGraph().getVertexByApplicationId(Integer.parseInt(e.getParm("applicationId").getValue().getContent())) != null) {
 
+            final Set<BusinessService> businessServices = m_stateMachine.getGraph().getVertices().stream().filter(v -> v.getBusinessService() != null).map(v -> v.getBusinessService()).collect(Collectors.toSet());
```

Comment:
The BusinessServiceGraph already has the knowledge of `application id -> vertex` and `reduction key -> vertex`. I'd rather see us leverage that indexing instead of having to iterate and search the whole graph again here.

Reply:
Yes, but I do not need the application or reduction key leaf node at this point. I need the business service that has the application edge or the reduction key edge attached.

Reply:
We can traverse the graph from there. My main concern here is performance, these searches will take a lot of cycles on large graphs.

Reply:
Since the graph is built from top-bottom, I only have one way to go down based on the business services to see which BS definitions are really affected. I can't get the business services for one leaf that are direct predecessors. Or am I missing something?

Reply:
We'll need to extend some interfaces to grab it from this context, but the complete graph is available. The state propagation goes upward, so it's possible to find the next affected BS from the leaf vertices.

Reply:
Ok. I see. I will change this...

Reply:
done